### PR TITLE
OpList and WorkerOp improvements and fixes

### PR DESCRIPTION
**Pull request by randomPoison**
I've made a number of improvements to `OpList` and `WorkerOp`, including some critical safety fixes that address some use-after-free issues that are currently present. Some variants of `WorkerOp` depend on data owned by the root `OpList`, but there were previously no lifetimes ensuring that a `WorkerOp` didn't outlive its `OpList`. I've added the appropriate lifetime markers to the various op types, and updated `OpList` and `WorkerOp` accordingly.

Additional minor improvements:

* `OpList` no longer allocates (previously done by creating a `Vec<WorkerOp>` internally), now it directly uses the from the underlying `Worker_OpList`.
* Added `OpList::iter` helper method in order to make `OpList` API more idiomatic.
* Added `OpList::len` added.

@davedissian If this is going to cause merge conflicts with the codegen changes, I'm fine to have this wait until after those are merged (assuming it's approved) and deal with the conflicts on my end.

### CODE REVIEW

File: spatialos-sdk/src/worker/commands.rs [hunk 1]
```diff
@@ -6,31 +6,31 @@ use spatialos_sdk_sys::worker::{
 };
 
 // TODO: Wrap Schema_CommandRequest
-pub struct CommandRequest {
+pub struct CommandRequest<'a> {
     pub component_id: ComponentId,
-    pub schema_type: *mut Schema_CommandRequest,
+    pub schema_type: &'a Schema_CommandRequest,
```
Comment (davedissian) [inline spatialos-sdk/src/worker/commands.rs]: Note that this is going to conflict pretty hard with my branch :P

Comment (randomPoison) [reply spatialos-sdk/src/worker/commands.rs]: I figured as much. I'm down to address conflicts once your changes get merged in :+1:

File: spatialos-sdk/src/worker/component.rs [hunk 2]
```diff
@@ -52,17 +52,23 @@ where
...
+    pub struct ComponentData<'a> {
         pub component_id: ComponentId,
         pub schema_type: SchemaComponentData,
-        pub user_handle: *mut Worker_ComponentDataHandle,
+        pub user_handle: *const Worker_ComponentDataHandle,
+
+        // NOTE: `user_handle` is borrowing data owned by the parent object, but it's a
+        // type-erased pointer that may be null, so we just mark that we're borrowing
+        // *something*.
+        pub _marker: PhantomData<&'a ()>,
```
Comment (jamiebrynes7) [inline spatialos-sdk/src/worker/component.rs]: Did not know you could do this. Pretty neat

File: spatialos-sdk/src/worker/op.rs [hunk 3]
```diff
@@ -16,40 +16,81 @@ use crate::worker::internal::utils::*;
...
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use spatialos_sdk::worker::connection::*;
+    /// # let connection: WorkerConnection = unimplemented!();
+    /// for op in connection.get_op_list(0).iter() {
+    ///     // Process `op`.
+    /// }
+    /// ```
+    pub fn iter(&self) -> Iter<'_> {
```
Comment (jamiebrynes7) [inline spatialos-sdk/src/worker/op.rs]: qq - What does `Iter<'_>` mean

File: spatialos-sdk/src/worker/op.rs [hunk 4]
```diff
@@ -462,17 +502,17 @@ pub struct EntityQueryResponseOp {
...
+    component_data: component::internal::ComponentData<'a>,
 }
 
-impl AddComponentOp {
+impl<'a> AddComponentOp<'a> {
     pub fn get<C: Component>(&self) -> Option<&C> {
         if C::ID == self.component_data.component_id {
             // TODO: Deserialize schema_type if user_handle is null.
-            Some(unsafe { &*(self.component_data.user_handle as *const _) })
+            Some(unsafe { &*(self.component_data.user_handle as *const _ as *const _) })
```
Comment (jamiebrynes7) [inline spatialos-sdk/src/worker/op.rs]: Is the double cast necessary?

File: spatialos-sdk/src/worker/component.rs (hunk 2 above)
Comment (randomPoison) [reply spatialos-sdk/src/worker/component.rs]: Yup! You actually have to use `PhantomData` if you have a generic parameter that you're not using in the type itself (as is the case here).

File: spatialos-sdk/src/worker/op.rs (hunk 3 above)
Comment (randomPoison) [reply spatialos-sdk/src/worker/op.rs]: It's new in the 2018 edition. If I recall correctly, you have to specify a lifetime when referencing a type that has a lifetime parameter, i.e. I couldn't do `fn iter(&self) -> Iter`, I'd have to do `fn iter<'a>(&'a self) -> Iter<'a>`. Doing `'_` says "the value I'm returning borrows some data, but I'm letting the compiler infer the lifetime of that data". The idea is that it makes it clear that the returned value borrows data, without all the boilerplate of manually writing out the lifetimes.

File: spatialos-sdk/src/worker/op.rs (hunk 4 above)
Comment (randomPoison) [reply spatialos-sdk/src/worker/op.rs]: Nope! It was briefly necessary in a previous iteration, but I changed things such that it's no longer necessary. Looks like I missed one when I was cleaning it up.

**Comment by randomPoison** [comment]
Updated!

**Comment by jamiebrynes7** [comment]
Same deal here - reopen PR on the repo directly and I'll reapprove 👍

**Pull request closed without merging**